### PR TITLE
gh-69929: Add more specific definition of \w

### PR DESCRIPTION
**Pull request by slateny**
#69929, #82747

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
@SnoopJeDi on #82747 suggested referring to the `str.isalnum()` documentation. I can confirm that that's correct: they both map to the `Py_UNICODE_ISALNUM` macro. I think it's also clearer, so could you do that instead?

**Comment by slateny** [comment]
Thanks, that's a good suggestion - do you know whether the wording can be more exact, so instead of `this includes most alphanumeric characters` it can be `this includes all alphanumeric characters`?

**Comment by slateny** [comment]
Thanks, that's a good suggestion - do you know whether the wording can be more exact, so instead of `this includes most alphanumeric characters` it can be `this includes all alphanumeric characters`?

File: Doc/library/re.rst [hunk 1]
```diff
@@ -523,10 +523,11 @@ character ``'$'``.
...
 ``\w``
    For Unicode (str) patterns:
-      Matches Unicode word characters; this includes most characters
-      that can be part of a word in any language, as well as numbers and
-      the underscore. If the :const:`ASCII` flag is used, only
-      ``[a-zA-Z0-9_]`` is matched.
+      Matches Unicode word characters; this includes most alphanumeric characters
+      as well as the underscore. In Unicode, alphanumeric characters are defined
+      to be the general categories L + N
+      (see `Unicode General Category Values <https://unicode.org/reports/tr44/#General_Category_Values>`_).
```
Comment (JelleZijlstra) [inline Doc/library/re.rst]:
```suggestion
      Matches Unicode word characters; this includes alphanumeric characters (as defined by :meth:`str.isalnum`)
      as well as the underscore (``_``).
```
Concrete wording suggestion

**Comment by zackw** [comment]
I like the idea of referring to the definition of `str.isalnum` for what regex \w means, but the definition of `str.isalnum` could also stand to be improved.  Right now it says

>  A character c is alphanumeric if one of the following returns True: c.isalpha(), c.isdecimal(), c.isdigit(), or c.isnumeric()

The definitions of isalpha, isdecimal, etc. refer to various Unicode properties and it's not obvious what the aggregate adds up to.  I'd like to see something like this added:

> Together, this includes everything in Unicode general categories L* and N*, plus U+005F (underscore).

**Comment by slateny** [comment]
I think a change like that would be better in a separate PR, so I've opened https://github.com/python/cpython/issues/96984 for possible discussion.

**Review by hauntsaninja — APPROVED** [review]

**Pull request merged**